src/features/soft_navigations/instrument/index.js

```diff
@@ -0,0 +1,66 @@
+import { originals } from '../../../common/config/config'
+import { isBrowserScope } from '../../../common/constants/runtime'
+import { handle } from '../../../common/event-emitter/handle'
+import { windowAddEventListener } from '../../../common/event-listener/event-listener-opts'
+import { now } from '../../../common/timing/now'
+import { debounce } from '../../../common/util/invoke'
+import { wrapEvents, wrapHistory } from '../../../common/wrap'
+import { InstrumentBase } from '../../utils/instrument-base'
+import { FEATURE_NAME, INTERACTION_TRIGGERS } from '../constants'
+
+/** The minimal time after a UI event for which no further events will be processed - i.e. a throttling rate to reduce spam.
+ * This also give some time for the new interaction to complete without being discarded by a subsequent UI event and wrongly attributed.
+ * This value is still subject to change and critique, as it is derived from beyond worst case time to next frame of a page.
+ */
+const UI_WAIT_INTERVAL = 1 / 10 * 1000 // assume 10 fps
+
+export class Instrument extends InstrumentBase {
+  static featureName = FEATURE_NAME
+  constructor (agentIdentifier, aggregator, auto = true) {
+    super(agentIdentifier, aggregator, FEATURE_NAME, auto)
+    if (!isBrowserScope || !originals.MO) return // soft navigations is not supported outside web env or browsers without the mutation observer API
+
+    const historyEE = wrapHistory(this.ee)
+    const eventsEE = wrapEvents(this.ee)
+
+    const trackURLChange = () => handle('newURL', [now(), '' + window.location], undefined, this.featureName, this.ee)
+    historyEE.on('pushState-end', trackURLChange)
+    historyEE.on('replaceState-end', trackURLChange)
+
+    try {
+      this.removeOnAbort = new AbortController()
+    } catch (e) {}
+    const trackURLChangeEvent = (evt) => handle('newURL', [evt.timeStamp, '' + window.location], undefined, this.featureName, this.ee)
+    windowAddEventListener('popstate', trackURLChangeEvent, true, this.removeOnAbort?.signal)
+
+    let oncePerFrame = false // attempt to reduce dom noice since the observer runs very frequently with below options
```

Comment:
running more frequently than expected also might be a consequence of calling .observe multiple times (every time it debounces)??

Reply:
The cb is executed per change or group of changes and there's no guarantee that all of those records are batched together per frame or per any other set interval. Calling observe on the same observer doesn't do that either.

> If you call observe() on a node that's already being observed by the same MutationObserver, all existing observers are automatically removed from all targets being observed before the new observer is activated.
> If the same MutationObserver is not already in use on the target, then the existing observers are left alone and the new one is added.